switch_mod/project/dispatch.py

```diff
@@ -160,6 +160,22 @@ def init_dispatch_timepoints(m):
     mod.PROJ_DISPATCH_POINTS = Set(
         dimen=2,
         initialize=init_dispatch_timepoints)
+    # make set array showing all timepoints when each project is active
+    # note: it might be cleaner to first define PROJ_ACTIVE_PERIODS, then use that
+    # to define PROJ_ACTIVE_TIMEPOINTS, PROJ_DISPATCH_POINTS, etc.
+    def proj_active_timepoints_rule(m, p):
+        if not hasattr(m, 'proj_active_timepoints_dict'):
+            d = m.proj_active_timepoints_dict = dict()
+            for p, t in m.PROJ_DISPATCH_POINTS:
+                if p not in d:
+                    d[p] = []
+                d[p].append(t)
+        # note: we could use pop(p) here to reduce memory use, but for some reason pyomo does 
+        # one extra call at the start, so that doesn't work.
+        # note: we return an empty list if a project has no active timepoints, usually 
+        # for projects that retire before the study
+        return m.proj_active_timepoints_dict.get(p, [])
```

Comment:
Move notes like the empty list one into the documentation at the beginning of define_components, and possibly in the doc field of the component itself. 
If you want to really clean up, you could do something like this:
```python
        # Use pop and delattr to gradually clear out and delete the cache
        active_timepoints = m.proj_active_timepoints_dict.pop(p)
        if len(m.proj_active_timepoints_dict) == 0:
        	delattr(m, 'proj_active_timepoints_dict')
        return active_timepoints
```